common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -22,64 +22,490 @@
+     * This method should only be called within a callback upon creating a successful YubiKey device connection.
+     * @param slot A PIV slot from which to read the certificate.
+     * @param piv A PivSession created from a SmartCardConnection that can interact with certificates located in the PIV slots on the YubiKey.
+     * @param certList A List collecting the YubiKitCertDetails of the certificates found on the YubiKey.
+     * @throws IOException          in case of connection error
+     * @throws ApduException        in case of an error response from the YubiKey
+     * @throws BadResponseException in case of incorrect YubiKey response
+     */
+    private void getAndPutCertDetailsInList(@NonNull final Slot slot, @NonNull final PivSession piv, @NonNull final List<YubiKitCertDetails> certList) throws IOException, ApduException, BadResponseException {
+        final String methodTag = TAG + ":getAndPutCertDetailsInList";
+        try {
+            final X509Certificate cert =  piv.getCertificate(slot);
+            //If there are no exceptions, add details of this cert to our certList.
+            certList.add(new YubiKitCertDetails(cert.getIssuerDN().getName(), cert.getSubjectDN().getName(), slot));
+        } catch (final ApduException e) {
+            //If sw is 0x6a82 (27266), This is a FILE_NOT_FOUND error, which we should ignore since this means the slot is merely empty.
+            if (e.getSw() == 0x6a82) {
+                Logger.verbose(methodTag, slot + " slot is empty.");
+            } else {
+                throw e;
+            }
+        }
+    }
+
+    /**
+     * Build and show Smartcard certificate picker, which also handles the rest of the smartcard CBA flow.
+     * @param certList A List holding the YubiKitCertDetails of the certificates found on the YubiKey.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     */
+    private void showSmartcardCertPickerDialog(@NonNull final List<YubiKitCertDetails> certList, @NonNull final ClientCertRequest request) {
+        final SmartcardCertPickerDialog certPickerDialog = new SmartcardCertPickerDialog(certList, mActivity);
+        certPickerDialog.setPositiveButtonListener(getSmartcardCertPickerDialogPositiveButtonListener(request));
+        certPickerDialog.setNegativeButtonListener(new SmartcardCertPickerDialog.NegativeButtonListener() {
+            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+            @Override
+            public void onClick() {
+                request.cancel();
+                synchronized (smartcardDialogLock) {
+                    mCurrentDialog = null;
+                }
+            }
+        });
+        certPickerDialog.setCancelCbaCallback(new SmartcardCertPickerDialog.CancelCbaCallback() {
+            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+            @Override
+            public void onCancel() {
+                request.cancel();
```

Comment:
Do we need to set `mCurrentDialog` to `null` here?

Reply:
I don't think so, as whenever we call `mCurrentDialog.onCancelCba()`, which calls the `CancelCbaCallback`, `mCurrentDialog` is immediately set to null afterwards